drivers/tee/optee/optee.c

```diff
@@ -177,9 +197,304 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
 	*reg1 = val;
 }
 
-static void handle_rpc_call(const struct device *dev, struct optee_rpc_param *param)
+static inline bool check_param_input(struct optee_msg_arg *arg)
+{
+	return (arg->num_params != 1 ||
+		arg->params[0].attr != OPTEE_MSG_ATTR_TYPE_VALUE_INPUT);
+}
+
+static void *optee_construct_page_list(void *buf, uint32_t len, uint64_t *phys_buf);
+
+static void handle_cmd_alloc(const struct device *dev, struct optee_msg_arg *arg,
+			     void **pages)
 {
+	int rc;
 	struct tee_shm *shm = NULL;
+	void *pl;
+	uint64_t pl_phys_and_offset;
+
+	arg->ret_origin = TEEC_ORIGIN_COMMS;
+
+	if (check_param_input(arg)) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	/*
+	 * TODO: OPTEE_RPC_SHM_TYPE_KERNEL and OPTEE_SHM_TYPE_APPL should be handled
```

Comment:
as far as I see you have already handled `OPTEE_RPC_SHM_TYPE_KERNEL` here and in `handle_cmd_free` function

Reply:
I left this comment to explicitly note that this expects different behavior for SHM types. I'll rephrase.

Reply:
There is another line it this todo. Saying that type_kernel is implemented

Reply:
better to delete it from todo and delete below line
just leave the comment that it is needed to handle `OPTEE_SHM_TYPE_APPL` only